src/Language/PureScript/TypeChecker/Deriving.hs

```diff
@@ -0,0 +1,150 @@
+-- Copyright   :  (c) Gershom Bazerman 2015
+-- License     :  MIT
+--
+-- Maintainer  :  Phil Freeman <paf31@cantab.net>
+-- Stability   :  experimental
+-- Portability :
+--
+-- |
+-- This module implements the generic deriving elaboration that takes place during typechecking.
+--
+-----------------------------------------------------------------------------
+
+{-# LANGUAGE FlexibleContexts, MultiParamTypeClasses, OverloadedStrings, PatternSynonyms #-}
+
+module Language.PureScript.TypeChecker.Deriving (
+    elaborateInstance
+) where
+
+import Data.List
+import Data.Maybe (fromMaybe)
+import Data.Ord (comparing)
+import qualified Data.Map as M
+
+import Control.Applicative
+import Control.Monad.State
+import Control.Monad.Error.Class (MonadError(..))
+
+import Language.PureScript.AST
+import Language.PureScript.Environment
+import Language.PureScript.Errors
+import Language.PureScript.Names
+import Language.PureScript.TypeChecker.Monad
+import Language.PureScript.Types
+
+elaborateInstance :: (Functor m, MonadState CheckState m, MonadError MultipleErrors m) => Expr -> m Expr
+elaborateInstance (TypeClassInstanceMemberFunction funName className t) =
+    fromMaybe noInstanceError $ lookup (unQualify className, funName) derivingList
+        where noInstanceError = throwError $ MultipleErrors [ErrorInInstance className [TypeConstructor t] (SimpleErrorWrapper $ MissingClassMember funName)]
+              derivingList = [((ProperName "Generic", Ident "toSpine"), mkSpineFunction t)
+                             ,((ProperName "Generic", Ident "fromSpine"), mkFromSpineFunction t)
+                             ,((ProperName "Generic", Ident "toSignature"), mkSignatureFunction t)]
+elaborateInstance e = return e
+
+mkSpineFunction :: (Functor m, MonadState CheckState m) => Qualified ProperName ->  m Expr
+mkSpineFunction t = do
+  ctors <- M.toList . dataConstructors <$> getEnv
+  tcs <- M.toList . typeClasses <$> getEnv
+  let prodConstructor = App (Constructor $ findName "SProd" ctors)
+      recordConstructor = App (Constructor $ findName "SRecord" ctors)
+      genericModule = case findName "Generic" tcs of (Qualified mn _) -> mn
+
+      mkCtorClause (ctorName, (_,_,typ,idents)) = CaseAlternative [ConstructorBinder ctorName (map VarBinder idents)] (Right caseResult)
+          where caseResult = App (prodConstructor (StringLiteral . runProperName . unQualify $ ctorName)) . ArrayLiteral $ zipWith toSpineFun (map (Var . Qualified Nothing) idents) (argTypes typ)
+
+      toSpineFun :: Expr -> Type -> Expr
+      toSpineFun i (ObjectType rec) =
+          lam "q" . recordConstructor . ArrayLiteral .
+              map (\(str,typ) -> ObjectLiteral [("recLabel", StringLiteral str), ("recValue", toSpineFun (Accessor str i) typ)])
+              $ decomposeRec rec
+      toSpineFun i _ = lam "q" $ App (mkVarMn genericModule "toSpine") i
+
+  return . lamCase "x" . map mkCtorClause . constructorsForType t $ ctors

```

Comment:
`x` is used, so we can't use `NullBinder` to avoid shadowing warnings, but maybe we can use a name which is valid JS but not valid PS, so nobody can use the same identifier for a top-level name. Dollar symbols are the usual hack here.